rp2040-hal/examples/i2c_async_irq.rs

```diff
@@ -0,0 +1,143 @@
+//! # I²C Example
+//!
+//! This application demonstrates how to talk to I²C devices with an RP2040.
+//! in an Async environment.
+//!
+//! This example sets a `waker_setter` and is able to sleep when the task is pending on i2c
+//! updates..
+//!
+//! It may need to be adapted to your particular board layout and/or pin assignment.
+//!
+//! See the `Cargo.toml` file for Copyright and license details.
+
+#![no_std]
+#![no_main]
+
+use core::{cell::RefCell, task::Waker};
+
+use critical_section::Mutex;
+
+// Ensure we halt the program on panic (if we don't mention this crate it won't
+// be linked)
+use panic_halt as _;
+
+// Alias for our HAL crate
+use rp2040_hal as hal;
+
+// Some traits we need
+use hal::fugit::RateExtU32;
+
+// Import required types & traits.
+use eh_async_1_0_alpha::i2c::I2c;
+use hal::{
+    gpio::{FunctionI2C, Pin, PullUp},
+    pac::{self, interrupt},
+    Clock,
+};
+
+/// The linker will place this boot block at the start of our program image. We
+/// need this to help the ROM bootloader get our code up and running.
+/// Note: This boot block is not necessary when using a rp-hal based BSP
+/// as the BSPs already perform this step.
+#[link_section = ".boot2"]
+#[used]
+pub static BOOT2: [u8; 256] = rp2040_boot2::BOOT_LOADER_GENERIC_03H;
+
+/// External high-speed crystal on the Raspberry Pi Pico board is 12 MHz. Adjust
+/// if your board has a different frequency
+const XTAL_FREQ_HZ: u32 = 12_000_000u32;
+
+/// Mutex holding an optional waker for the i2c peripiheral.
+static WAKER: Mutex<RefCell<Option<Waker>>> = Mutex::new(RefCell::new(None));
+
+/// Waker setter method updating the waker of the async task.
+pub fn i2c_waker_setter(waker: Waker) {
+    critical_section::with(|cs| {
+        *WAKER.borrow_ref_mut(cs) = Some(waker);
+    });
+}
+
+#[interrupt]
+fn I2C1_IRQ() {
+    let i2c1 = unsafe { &rp2040_hal::pac::Peripherals::steal().I2C1 };
+    i2c1.ic_intr_mask.modify(|_, w| {
+        w.m_tx_abrt()
+            .enabled()
+            .m_tx_empty()
+            .enabled()
+            .m_rx_full()
+            .enabled()
+    });
+    critical_section::with(|cs| WAKER.borrow_ref_mut(cs).take().map(Waker::wake));
```

Comment:
What does this do? Can we write it out as multiple lines with intermediate variable names and some comments?

Reply:
I think I'd also prefer `if let Some(waker) = borrowed_global_waker.take() { Wake::wake(waker) }` or similar, even if clippy might say "yeah, but you can just use .map(Waker::wake)", because it's non-obvious where why you would grab a global and then map it to a different type, and then drop the value.